usersession/userd/launcher.go

```diff
@@ -123,6 +130,157 @@ func (s *Launcher) OpenURL(addr string, sender dbus.Sender) *dbus.Error {
 	return nil
 }
 
+// OpenDesktopEntryEnv implements the 'OpenDesktopEntryEnv' method of the 'io.snapcraft.Launcher'
+// DBus interface. The desktop_file_id is described here:
+// https://standards.freedesktop.org/desktop-entry-spec/desktop-entry-spec-latest.html#desktop-file-id
+func (s *Launcher) OpenDesktopEntryEnv(desktop_file_id string, env []string, sender dbus.Sender) *dbus.Error {
+	desktop_file, err := desktopFileIdToFilename(existsOnFileSystem, desktop_file_id)
+	if err != nil {
+		return dbus.MakeFailedError(err)
+	}
+
+	exec_command, err := readExecCommandFromDesktopFile(desktop_file)
+	if err != nil {
+		return dbus.MakeFailedError(err)
+	}
+
+	args, err := parseExecCommand(exec_command)
+	if err != nil {
+		return dbus.MakeFailedError(err)
+	}
+
+	// Before rejoining args to create a new command-line: escape "\" escapes; escape "\"" quotes; and, wrap each arg in "\"" quotes
+	for i := 0; i != len(args); i++ {
+		args[i] = strings.Replace(args[i], "\\", "\\\\", -1)
+		args[i] = strings.Replace(args[i], "\"", "\\\"", -1)
+		args[i] = "\"" + args[i] + "\""
+	}
+
+	// Invoke indirectly via sh to unparent and avoid potentially leaving a zombie
+	cmd := exec.Command("sh", "-c", strings.Join(args, " ")+"&")
+	cmd.Env = os.Environ()
+	for _, e := range env {
+		if !strutil.ListContains(allowedEnvVars, strings.SplitN(e, "=", 2)[0]) {
+			return dbus.MakeFailedError(fmt.Errorf("Supplied environment variable %q is not allowed", e))
+		}
+
+		cmd.Env = append(cmd.Env, e)
+	}
+
+	if cmd.Run() != nil {
+		return dbus.MakeFailedError(fmt.Errorf("cannot run %q", exec_command))
+	}
+
+	return nil
+}
+
+type fileExists func(string) bool
+
+func existsOnFileSystem(desktop_file string) bool {
+	fileStat, err := os.Stat(desktop_file)
+
+	return err == nil && !fileStat.IsDir()
+}
+
+// findDesktopFile recursively tries each subdirectory that can be formed from the (split) desktop file ID.
+func findDesktopFile(desktop_file_exists fileExists, base_dir string, splitFileId []string) *string {
```

Comment:
If it's snaps starting snaps, then there's no need to recurse into subdirectories, all desktop files should be under `dirs.SnapDesktopFilesDir`.

Reply:
I was unaware of that guarantee for the current usecase.

We're likely to encounter desktop entry trees without this constraint sooner rather than later as looking in XDG_DATA_DIRS is something that would make sense for the "confined desktop" that's being talked about.

This code works for both, I don't see the value in changing it to rely on an assumption that could prove short lived.